fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/Election.java

```diff
@@ -266,50 +269,126 @@ public static String generateElectionQuestionId(String electionId, String questi
    */
   public static String generateElectionVoteId(
-      String electionId,
-      String questionId,
-      List<Integer> voteIndex,
-      String writeIn,
-      boolean writeInEnabled) {
+          String electionId,
+          String questionId,
+          Integer voteIndex,
+          String writeIn,
+          boolean writeInEnabled) {
     // If write_in is enabled the id is formed with the write_in string
     // If write_in is not enabled the id is formed with the vote indexes (formatted as int1, int2,
     // ). The vote are concatenated and brackets are removed from the array toString representation
-    String voteIndexFormat = voteIndex
-            .toString()
-            .replace("]", "")
-            .replace("[", "");
     return Hash.hash(
-            "Vote", electionId, questionId, writeInEnabled ? writeIn : voteIndexFormat);
+        "Vote", electionId, questionId, writeInEnabled ? writeIn : voteIndex.toString());
+  }
+
+  /**
+   * Generate the id for a vote of dataCastVote.
+   * https://github.com/dedis/popstellar/blob/master/protocol/query/method/message/data/dataCastVote.json
+   *
+   * @param electionId ID of the Election
+   * @param questionId ID of the Election question
+   * @param voteIndex index(es) of the vote
+   * @param writeInEncrypted string representing the write in
+   * @param writeInEnabled boolean representing if write enabled or not
+   * @return the ID of an election question computed as
+   *     Hash('Vote'||election_id||question_id||(encrypted_vote_index(es)|encrypted_write_in))
+   */
+  public static String generateEncryptedElectionVoteId(
+          String electionId,
+          String questionId,
+          String voteIndex,
+          String writeInEncrypted,
+          boolean writeInEnabled) {
+    // HashLen('Vote', election_id, question_id, (encrypted_vote_index(es)|encrypted_write_in))),
+    // concatenate vote indexes - must sort in alphabetical order and use delimiter ','"
+    return Hash.hash(
+            "Vote", electionId, questionId, writeInEnabled ? writeInEncrypted : voteIndex);
+  }
+
+  /**
+   * Computes the hash for the registered votes, when terminating an election (sorted by message
+   * id's alphabetical order)
+   *
+   * @return the hash of all registered votes
+   */
+  public String computerRegisteredVotes() {
+    List<String> listOfVoteIds = new ArrayList<>();
+    // Since messageMap is a TreeMap, votes will already be sorted in the alphabetical order of
+    // messageIds
+    for (PublicKey senderPk : messageMap.values()) {
+      if (getElectionVersion() == ElectionVersion.OPEN_BALLOT) {
+        for (ElectionVote vote : openVoteByPublicKey.get(senderPk)) {
+          listOfVoteIds.add(vote.getId());
+        }
+      } else {
+        for (ElectionEncryptedVote vote : encryptedVoteByPublicKey.get(senderPk)) {
+          listOfVoteIds.add(vote.getId());
+        }
+      }
+    }
```

Comment:
There is a bit of code duplication here.

I'm certain you can find a way to modularize this.

Reply:
Hmm, i don't know how to achieved, i could write two functions but I dont know if thats very useful